File: tools/pplx-exporter/src/audit.ts
import { appendFileSync, existsSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";
import type { AuditEntry } from "./types.js";

export function appendAudit(path: string, entry: AuditEntry): void {
  const dir = dirname(path);
  if (!existsSync(dir)) mkdirSync(dir, { recursive: true });

  const line = JSON.stringify({ ...entry, ts: entry.ts ?? new Date().toISOString() }) + "\n";
  appendFileSync(path, line, "utf-8");
}

export function auditThreadFetched(path: string, slug: string, bytes: number, hash: string): void {
  appendAudit(path, {
    ts: new Date().toISOString(),
    event: "thread_fetched",
    slug,
    bytes,
    hash,
    status: "ok",
  });
}

export function auditThreadFailed(path: string, slug: string, error: string, retry: number): void {
  appendAudit(path, {
    ts: new Date().toISOString(),
    event: "thread_fetch_failed",
    slug,
    error,
    retry,
    status: "error",
  });
}

export function auditDiscoveryComplete(path: string, totalThreads: number): void {
  appendAudit(path, {
    ts: new Date().toISOString(),
    event: "discovery_complete",
    status: "ok",
    bytes: totalThreads,
  });
}

export function auditRunStart(path: string, runId: string): void {
  appendAudit(path, {
    ts: new Date().toISOString(),
    event: "run_start",
    status: "ok",
    slug: runId,
  });
}

export function auditRunEnd(path: string, totalExported: number): void {
  appendAudit(path, {
    ts: new Date().toISOString(),
    event: "run_end",
    status: "ok",
    bytes: totalExported,
  });
}

File: tools/pplx-exporter/pplx_exporter/audit.py
"""Audit log — JSONL append-only accountability trail."""

from __future__ import annotations

import json
from datetime import UTC, datetime
from pathlib import Path


def _now() -> str:
    return datetime.now(UTC).isoformat()


def _append(path: Path, entry: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as f:
        f.write(json.dumps(entry, ensure_ascii=False) + "\n")


def audit_run_start(path: Path, run_id: str) -> None:
    _append(path, {"ts": _now(), "event": "run_start", "run_id": run_id})


def audit_run_end(path: Path, run_id: str, total: int) -> None:
    _append(path, {"ts": _now(), "event": "run_end", "run_id": run_id, "total_exported": total})


def audit_discovery_complete(path: Path, count: int) -> None:
    _append(path, {"ts": _now(), "event": "discovery_complete", "thread_count": count})


def audit_thread_fetched(path: Path, slug: str, byte_count: int, sha: str) -> None:
    _append(path, {"ts": _now(), "event": "thread_fetched", "slug": slug, "bytes": byte_count, "hash": sha})


def audit_thread_failed(path: Path, slug: str, error: str, retry: int) -> None:
    _append(path, {"ts": _now(), "event": "thread_failed", "slug": slug, "error": error, "retry": retry})
